src/coordination/coordinator_instance.cpp

```diff
@@ -214,78 +223,148 @@ auto CoordinatorInstance::ShowInstances() const -> std::vector<InstanceStatus> {
-      return {.instance_name = instance.InstanceName(),
-              .management_server = instance.ManagementSocketAddress(),  // show non-resolved IP
-              .bolt_server = instance.BoltSocketAddress(),              // show non-resolved IP
-              .cluster_role = stringify_repl_role(instance),
-              .health = stringify_repl_health(instance),
-              .last_succ_resp_ms = instance.LastSuccRespMs().count()};
-    };
+  std::ranges::transform(raft_state_->GetReplicationInstances(), std::back_inserter(instances_status),
+                         process_repl_instance_as_follower);
+  spdlog::trace("Processed all replication instances as follower, returning report as follower.");
+  return instances_status;
+}
 
-    {
-      auto lock = std::shared_lock{coord_instance_lock_};
-      std::ranges::transform(repl_instances_, std::back_inserter(instances_status), process_repl_instance_as_leader);
+auto CoordinatorInstance::ShowInstances() -> std::pair<ShowInstancesState, std::vector<InstanceStatus>> {
+  if (!is_leader_ready_) {
+    spdlog::trace("Processing show instances as follower {}", std::this_thread::get_id());
+    auto const leader_id = raft_state_->GetLeaderId();
+    if (leader_id == raft_state_->GetCoordinatorId()) {
+      spdlog::trace("Coordinator itself not yet leader, returning report as follower.");
+      return {ShowInstancesState::FOLLOWER,
+              GetAllInstancesStatusAsFollower()};  // We don't want to ask ourselves for instances, as coordinator is
+                                                   // not ready still as leader
     }
-  } else {
-    auto const stringify_inst_status =
-        [raft_state_ptr = raft_state_.get()](ReplicationInstanceState const &instance) -> std::string {
-      if (raft_state_ptr->IsCurrentMain(instance.config.instance_name)) {
-        return "main";
-      }
-      if (raft_state_ptr->HasMainState(instance.config.instance_name)) {
-        return "unknown";
+
+    if (leader_id == -1) {
+      spdlog::trace("No leader found, returning report as follower {}", std::this_thread::get_id());
+      return {ShowInstancesState::FOLLOWER, GetAllInstancesStatusAsFollower()};
+    }
+    CoordinatorInstanceConnector *follower{nullptr};
+    // TODO(antoniofilipovic) move this part to separate function
+    {
+      auto connectors = coordinator_connectors_.Lock();
+      if (!connectors->contains(leader_id)) {
+        auto all_coordinators = raft_state_->GetCoordinatorInstances();
+
+        auto coord = std::ranges::find_if(all_coordinators,
+                                          [leader_id](auto const &coord) { return coord.coordinator_id == leader_id; });
+
+        MG_ASSERT(coord != all_coordinators.end());
+        connectors->emplace(
+            leader_id, CoordinatorInstanceConnector{CoordinatorInstanceManagementServerConfig{coord->management_server},
```

Comment:
Really don't understand caching based on leader_id? I think you don't need it. Connectors can have coordinator id in them and you just need to retrieve appropriate connection and that's all.

Reply:
Agreed, that is from version before so I removed it